File: livraria-api/src/main/java/br/com/tpa/livrariaapi/service/ClientesService.java
package br.com.tpa.livrariaapi.service;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import br.com.tpa.livrariaapi.model.Cliente;
import br.com.tpa.livrariaapi.repository.ClientesRepository;

@Service
public class ClientesService {
	
	@Autowired
	private ClientesRepository clientesRepository;
	
	public Cliente incluir(Cliente cliente) {
		return clientesRepository.save(cliente);
	}
	
	public List<Cliente> listar() {
		return clientesRepository.findAll();
	}
	
	public Optional<Cliente> buscar(Long codigo) {
		Optional<Cliente> cliente = clientesRepository.findById(codigo);
		return cliente;
	}
	
	public void excluir(Cliente cliente) {
		clientesRepository.delete(cliente);
	}

	public List<Cliente> buscarPorNome(String nome) {
		return clientesRepository.buscarPorNome("%" + nome + "%");
	}
	
	public List<Cliente> buscarPorCpf(String cpf) {
		return clientesRepository.buscarPorCpf(cpf);
	}
}

File: livraria-ui/src/app/services/clientes.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Cliente } from '../Cliente';

@Injectable({
  providedIn: 'root'
})
export class ClientesService {

  private url = 'http://localhost:8080/clientes/';
  private clienteEditar: Cliente;

  constructor(private http: HttpClient) { }

  incluir(cliente: Cliente): Observable<any> {
    return this.http.post(this.url, cliente);
  }

  listar() {
    return this.http.get<any>(this.url);
  }

  buscar(codigo: string) {
    const params = new HttpParams().set('codigo', codigo);
    return this.http.get<any>(this.url, {params});
  }

  excluir(codigo: string) {
    this.http.delete(this.url + codigo)
    .subscribe();
  }

  clienteSelecionado(): Cliente {
    return this.clienteEditar || {} as Cliente;
  }

  prepararEditar(clienteEditar: Cliente) {
    this.clienteEditar = clienteEditar;
  }

  buscarPorNome(nome: string) {
    const params = new HttpParams().set('nome', nome);
    return this.http.get<any>(this.url, {params});
  }

  buscarPorCpf(cpf: string) {
    const params = new HttpParams().set('cpf', cpf);
    return this.http.get<any>(this.url, {params});
  }

}
